src/operator/join.ts

```diff
@@ -0,0 +1,172 @@
+import {Observable} from '../Observable';
+import {ArrayObservable} from '../observable/ArrayObservable';
+import {Operator} from '../Operator';
+import {Subscriber} from '../Subscriber';
+import {OuterSubscriber} from '../OuterSubscriber';
+import {InnerSubscriber} from '../InnerSubscriber';
+import {subscribeToResult} from '../util/subscribeToResult';
+import {take} from './take';
+
+/**
+ * Returns an Observable that combines the items emitted by two Observables, and selects which items to combine
+ * based on duration-windows that you define on a per-item basis via the durtionSelector functions. You implement
+ * these windows as Observables whose lifespans begin with each item emitted by either Observable. When such a
+ * window-defining Observable either emits an item or completes, the window for the item it is associated with
+ * closes. So long as an item's window is open, it will combine with any item emitted by the other Observable.
+ * `resultSelector` is the function you define to combine the items.
+ *
+ * @param {Observable} right  the right observable sequence to join elements for
+ * @param {Function} leftDurationSelector  a function to select the duration (expressed as an observable sequence)
+ * of each element of the left observable sequence, used to determine overlap
+ * @param {Function} rightDurationSelector  a function to select the duration (expressed as an observable sequence)
+ * of each element of the left observable sequence, used to determine overlap
+ * @param {Function} resultSelector  a function invoked to compute a result element for any two overlapping elements
+ * of the left and right observable sequences.
+ * @returns {Observable} an observable sequence that contains result elements computed from source elements that
+ * have an overlapping duration
+ */
+export function join<T1, T2, D1, D2, R>(right: Observable<T2>,
+                                        leftDurationSelector: (value: T1) => Observable<D1>,
+                                        rightDurationSelector: (value: T2) => Observable<D2>,
```

Comment:
The type of the duration selector observables should be `any`, since the typing doesn't at all matter there. So we don't need D1 and D2, I don't think.